### PR TITLE
feat: Nis1 ImportanceTransfer transaction

### PR DESCRIPTION
**Pull request by rg911**
- ImportanceTransfer serialization and attaching signature
- Note: this will very likely be changed once we have the Nis1 catbuffer buffer.
- Structure created for future refactoring / abstractions

### CODE REVIEW

File: src/core/nis1/Nis1Transaction.ts [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+     */
+    public get signature(): Uint8Array {
+        return this.payload.subarray(-64);
+    }
+
+    /**
+     * Serialize transaction
+     * @returns - Transaction payload bytes
+     */
+    public serialize(): Uint8Array {
```
Comment (fboucquez) [inline src/core/nis1/Nis1Transaction.ts:82]: what's the difference between serialize and internal payload?

File: src/core/nis1/Nis1Transaction.ts [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+import { Duration, Instant } from '@js-joda/core';
+import { toBufferLE } from 'bigint-buffer';
+import { Key } from '../Key';
+import { Converter } from '../utils/Converter';
+import { Serializer } from './transactions/Serializer';
+
+/**
+ * Generic Nis1 transaction class
+ */
+export class Nis1Transaction<T extends Serializer> {
```
Comment (fboucquez) [inline src/core/nis1/Nis1Transaction.ts:27]: Has nis1 transaciton a transaction hash?

**Comment by fboucquez** [comment]
Is it possible to generate transaction vector tests like https://github.com/symbol/sdk-typescript/blob/main/test/vector-tests/resources/test-transaction-signature.json from running the original nis1 SDK unit tests or pulling transactions from a node?

File: src/core/nis1/Nis1Transaction.ts (hunk 2 above)
Comment (rg911) [reply src/core/nis1/Nis1Transaction.ts:27]: don't think it has generation hash, but transaction hash yes, but this PR is just following python sdk for now, (until we have catbuffer schema ready)

File: src/core/nis1/Nis1Transaction.ts (hunk 1 above)
Comment (rg911) [reply src/core/nis1/Nis1Transaction.ts:82]: the internal payload may have signature & signature length

**Comment by rg911** [comment]
> Is it possible to generate transaction vector tests like https://github.com/symbol/sdk-typescript/blob/main/test/vector-tests/resources/test-transaction-signature.json from running the original nis1 SDK unit tests or pulling transactions from a node?
> 
> I think we should focus on nis1 catbuffer. Unsure the design, if it's the same symbol schema extended or a new schema.
> 
> If we extend the same schema we could possibly share some types but we would need to prefix the schemas objects with Symbol or Nis1 when not shared. LIke `SymbolTransferTransactionBuilder`, `SymbolTransferTransactionBodyBuilder` and `Nis1SymbolTransactionBuilder`.
> 
> If we don't share the schema, we could create 2 catbuffer libraries like:
> 
> ```
>  "catbuffer-symbol": "^1.0.1-alpha-202108062013",
>  "catbuffer-nis1": "^1.0.1-alpha-202108062013",
> ```
> 
> Catbuffer builder names could be the same, the import would be different.

Agree, we will have a look at both symbol & nis1 generation once nis1 schema is ready.

**Comment by Jaguar0625** [comment]
generally, i'd prefer if SymbolDeadline => Deadline is separate review since it is mostly unrelated to adding new tx

**Review by Jaguar0625 — CHANGES_REQUESTED** [review]
please split out SymbolDeadline => Deadiline renaming into a separate review as we should try not to mix cleanup/renames and new features in a single PR

**Comment by rg911** [comment]
> please split out SymbolDeadline => Deadiline renaming into a separate review as we should try not to mix cleanup/renames and new features in a single PR

SymbolDeadline has been put back, and I have added a temp NIs1Deadeline class for this PR. will consolidate both in a new PR

**Comment by fboucquez** [comment]
> 
> SymbolDeadline has been put back, and I have added a temp NIs1Deadeline class for this PR. will consolidate both in a new PR

I reckon it should have been easier to create a new PR renaming/moving SymbolDeadline -> Deadline, merge that one to main and then update this pr.

File: src/core/nis1/Nis1Transaction.ts (hunk 2 above)
Comment (AnthonyLaw) [reply src/core/nis1/Nis1Transaction.ts:27]: nope, nis1 doesn't have a generation hash.

File: src/core/constants.ts [hunk 3]
```diff
@@ -14,17 +14,37 @@
...
 
+/**
+ * EPOCH
+ */
+export const SymbolMainnetEpochTime = Date.UTC(2021, 2, 16, 0, 6, 25);
+export const SymbolTestnetEpochTime = Date.UTC(2021, 2, 25, 17, 56, 17);
+export const SymbolMainnetEpoch = 1615853185;
+export const SymbolTestnetEpoch = 1616694977;
+export const Nis1MainnetEpochTime = Date.UTC(2015, 2, 29, 0, 6, 25);
+export const Nis1MainnetEpoch = 1425859585;
```
Comment (AnthonyLaw) [inline src/core/constants.ts:25]: maybe it should rename to Nis1Epoch?

File: src/core/nis1/Nis1Deadline.ts [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+import { ChronoUnit, Duration, Instant, LocalDateTime, ZoneId } from '@js-joda/core';
+/**
+ * The deadline of the transaction. The deadline is given as the number of seconds elapsed since the creation of the nemesis block.
+ * If a transaction does not get included in a block before the deadline is reached, it is deleted.
+ */
+export class Nis1Deadline {
+    /**
+     * Deadline value (without Nemesis epoch adjustment)
+     */
+    public adjustedValue: number;
```
Comment (Jaguar0625) [inline src/core/nis1/Nis1Deadline.ts:26]: is this ms or s?

File: src/core/nis1/Nis1Transaction.ts [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+     * @param keyPair - Signer key pair
+     */
+    public sign(keyPair: Nis1KeyPair): void {
+        this.signerPublicKey = keyPair.publicKey; // Re-assign signer's public key
+        const payload = this.serialize();
+        const signature = keyPair.sign(payload);
+
+        this.internalPayload = Buffer.alloc(4 + payload.length + 4 + signature.length)
+            .fill(Converter.numberToUint8(payload.length, 4), 0)
+            .fill(this.serialize(), 4)
```
Comment (Jaguar0625) [inline src/core/nis1/Nis1Transaction.ts:105]: payload?

**Review by Jaguar0625 — DISMISSED** [review]

File: src/core/nis1/Nis1Transaction.ts (hunk 5 above)
Comment (rg911) [reply src/core/nis1/Nis1Transaction.ts:105]: There's a public property to return the payload (from internalPayload), the internalPayload is supposed to be untouchable by the end user

File: src/core/nis1/Nis1Deadline.ts (hunk 4 above)
Comment (rg911) [reply src/core/nis1/Nis1Deadline.ts]: hmm, just noticed now Nis1 is using seconds and Symbol is using milliseconds, will need to refactor the Deadline class I think, in a different PR perhaps

**Review by Jaguar0625 — APPROVED** [review]

**Pull request merged**